R/enum.R

```diff
@@ -25,6 +25,108 @@ enum <- function(enumValues) {
   return(myEnum)
 }
 
+#' Get the key mapped to the given value in an \code{enum}
+#'
+#' @param enum The enum where the key-value pair is stored
+#' @param value The value that is mapped to the key
+#'
+#' @return Key under which the value is stored. If the value is not in the enum, \code{NULL} is returned
+#' @export
 getEnumKey <- function(enum, value) {
-  names(which(enum == value))
+  output <- names(which(enum == value))
+  if (length(output) == 0) {
+    return(NULL)
+  }
+  return(output)
+}
+
+#' Return the value that is stored under the given key. If the key is not present, an error is thrown.
+#'
+#' @param enum enum that contains the key-value pair
+#' @param key key under which the value is stored
+#'
+#' @return Value that is assigned to \code{key}
+#' @export
+enumGetValue <- function(enum, key) {
+  if (!enumHasKey(key, enum)) {
+    stop(messages$errorKeyNotInEnum(key))
+  }
+
+  return(enum[[key]])
+}
+
+#' Return all keys of an enum
+#'
+#' @param enum \code{enum} containing the keys
+#'
+#' @return List of key names
+#' @export
+enumKeys <- function(enum) {
+  names(enum)
+}
+
+#' Check if an enum has a certain key.
+#'
+#' @param key Key to check for
+#' @param enum Enum where to look for the key
+#'
+#' @return TRUE if a key-value pair for \code{key} exists, FALSE otherwise
+#' @export
+enumHasKey <- function(key, enum) {
+  return(any(enumKeys(enum) == key))
+}
+
+#' Add a new key-value pairs to an enum.
+#'
+#' @param keys Keys of the values to be added
+#' @param values Values to be added
+#' @param enum enum the key-value pairs should be added to.
+#' WARNING: the original object is not modified!
+#' @param overwrite if TRUE and a value with any of the given \code{keys} exists,
+#' it will be overwritten with the new value. Otherwise, an error is thrown. Default is FALSE.
+#'
+#' @return Enum with added key-value pair.
+#' @export
+#'
+#' @examples
+#' myEnum <- enum(c(a = "b"))
+#' myEnum <- enumPut("c", "d", myEnum)
+#' myEnum <- enumPut(c("c", "d", "g"), c(12, 2, "a"), myEnum, overwrite = TRUE)
+enumPut <- function(keys, values, enum, overwrite = FALSE) {
+  validateIsSameLength(keys, values)
+
+  for (i in seq_along(keys)) {
+    if (enumHasKey(keys[[i]], enum) && !overwrite) {
+      stop(messages$errorKeyInEnumPresent(keys[[i]]))
+    }
+    enum[[keys[[i]]]] <- values[[i]]
+  }
+
+  return(enum)
+}
+
+#' Remove an entry from the enum.
+#'
+#' @param keys Key(s) of entries to be removed from the enum
+#' @param enum Enum from which the entries to be removed
+#' WARNING: the original object is not modified!
+#'
+#' @return Enum without the removed entries
+#' @export
+enumRemove <- function(keys, enum) {

```

Comment:
woudl it work with keys being ONE value and not an array?

Reply:
Yepp.

Reply:
So why do we always do sthg like this in ospsuite-r

quantities<- c(quantities) to ensure that we have a list

Reply:
I think... Because a quantity (or any R6-object) by themselves are kind of a list, and trying to iterate over an R6-object produces some mess... Having an object as enum-key would not work.

Reply:
oh yeah that's right. Thanks brother